src/main/java/org/spongepowered/api/text/serializer/TextTemplateConfigSerializer.java

```diff
@@ -0,0 +1,129 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.serializer;
+
+import com.google.common.reflect.TypeToken;
+import ninja.leaping.configurate.ConfigurationNode;
+import ninja.leaping.configurate.objectmapping.ObjectMappingException;
+import ninja.leaping.configurate.objectmapping.serialize.TypeSerializer;
+import org.spongepowered.api.text.LiteralText;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.TextTemplate;
+import org.spongepowered.api.text.format.TextFormat;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Represents a {@link TypeSerializer} for {@link TextTemplate}s. TextTemplates
+ * are serialized in two parts.
+ *
+ * <p>First, the template's arguments as defined by
+ * {@link TextTemplate#getArguments()} are serialized to the "arguments" node.
+ * This is where the argument definitions are kept.</p>
+ *
+ * <p>Second, the template's text representation as defined by
+ * {@link TextTemplate#toText()} is serialized to the "content" node.</p>
+ *
+ * <p>Deserialization is a bit more complicated. We start by loading the
+ * "content" Text and check the root Text element as well as it's children. If
+ * a {@link LiteralText} value is found that is wrapped in curly braces we
+ * check to see if the value inside the braces is defined as an argument in the
+ * "arguments" nodes. If so, we use the name and format from the original
+ * LiteralText and obtain whether the argument is optional from the definition.
+ * This is repeated until there are no more Text children to check and we
+ * return a TextTemplate of the elements we have collected.</p>
+ */
+public class TextTemplateConfigSerializer implements TypeSerializer<TextTemplate> {
+
+    private static final String NODE_CONTENT = "content";
+    private static final String NODE_ARGS = "arguments";
+    private static final String NODE_OPT = "optional";
+    private static final String NODE_DEF_VAL = "defaultValue";
+    private static final String NODE_OPTIONS = "options";
+    private static final String NODE_OPEN_ARG = "openArg";
+    private static final String NODE_CLOSE_ARG = "closeArg";
+
+    @SuppressWarnings("NullableProblems") private ConfigurationNode root;
```

Comment:
Suppresses the warning that it has to be initialized, which it is by deserialized, but that's not recognized. Could make it Nullable but then it would complain if you don't check if it's not null which is unnecessary.